### PR TITLE
Make fatal warnings not fail compilation early & aggregate warns

### PR DESCRIPTION
**Pull request by szymon-rd**
Final PR. Adds functionality that changes the behaviour of fatal-warnings - fixes https://github.com/lampepfl/dotty/issues/18634
PR 5/5 (merge consecutively, per [Nicolas' suggestion](https://github.com/lampepfl/dotty/pull/18829#pullrequestreview-1759527204)

Split up version of https://github.com/lampepfl/dotty/pull/18829

### CODE REVIEW

**Comment by szymon-rd** [comment]
There was some problem with moving these tests to warn when I tried it before but I will look a bit into that now to make sure.

**Comment by szymon-rd** [comment]
I moved most of them to warn

File: compiler/src/dotty/tools/dotc/reporting/Reporter.scala [hunk 1]
```diff
@@ -214,6 +209,10 @@ abstract class Reporter extends interfaces.ReporterResult {
   def incomplete(dia: Diagnostic)(using Context): Unit =
     incompleteHandler(dia, ctx)
 
+  def finalizeReporting()(using Context) = 
+    if (hasWarnings && ctx.settings.XfatalWarnings.value)
+      report(new Error("No warnings can be incurred under -Werror.", NoSourcePosition))
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/reporting/Reporter.scala:214]:
```suggestion
      report(new Error("No warnings can be incurred under -Werror (or -Xfatal-warnings)", NoSourcePosition))
```

File: compiler/test/dotty/tools/dotc/CompilationTests.scala [hunk 2]
```diff
@@ -216,7 +216,6 @@ class CompilationTests {
   @Test def checkInitGlobal: Unit = {
     implicit val testGroup: TestGroup = TestGroup("checkInitGlobal")
     val options = defaultOptions.and("-Ysafe-init-global", "-Xfatal-warnings")
```
Comment (nicolasstucki) [inline compiler/test/dotty/tools/dotc/CompilationTests.scala:218]: We should inline `options` into `compileFilesInDir("tests/init-global/pos", options,`

**Review by nicolasstucki — COMMENTED** [review]
Otherwise LGTM

**Review by nicolasstucki — APPROVED** [review]

**Comment by nicolasstucki** [comment]
This uncovered a bug in `tests/neg/inline-unstable-accessors.scala`. This should not be fixed in this PR. I opened issue #19572 to track this bug.

To workaround this issue do 
```diff
package object H:
  inline def inlined =
-   G.valBinaryAPI2 +
+   // G.valBinaryAPI2 + // FIXME should error (now fails -Ycheck)
    G.valBinaryAPI3
```

```diff
package J:
  inline def inlined =
-    I.valBinaryAPI2 +
+   //  I.valBinaryAPI2 + // FIXME should error (now fails -Ycheck)
    I.valBinaryAPI3
```

and update the checkfile.

**Pull request merged**